torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,724 @@
+    path_prefixes_ = py::module::import("torch.profiler.python_tracer")
+        .attr("_prefix_regex")().cast<std::string>();
+
+    module_call_code_ = py::module::import("torch.nn")
+        .attr("Module")
+        .attr("__call__")
+        .attr("__code__")
+        .ptr();
+}
+
+void PythonTracer::start(size_t max_threads) {
+    TORCH_CHECK(!active_, "PythonTracer is already active")
+    TORCH_CHECK(!trace_contexts_.size(), "PythonTracer should not have active contexts");
+    TORCH_CHECK(max_threads > 0, "max_threads must be positive, got ", max_threads);
+    TORCH_CHECK(
+        max_threads <= max_py_threads,
+        "max_threads must be less than or equal to ", max_py_threads);
+
+    pybind11::gil_scoped_acquire gil;
+    auto t0 = now();
+
+    // Loop over all threads within the current interpreter. We will need to
+    // register a trace function with each thread. We set the current thread to
+    // position zero to ensure that it is traced, and so we can restore the
+    // thread state after registration.
+    std::vector<PyThreadState*> thread_states { PyThreadState_Get() };
+    if (max_threads > 1) {
+        auto thread_state = thread_states[0];
+        while (thread_state != nullptr) {
+            if (thread_state != thread_states[0]) {
+                thread_states.push_back(thread_state);
+            }
+            thread_state = PyThreadState_Next(thread_state);
+        }
+
+        if (thread_states.size() > max_threads) {
+            std::cout << "Warning: can only trace " << max_threads << " threads. "
+                    << thread_states.size() << " are currently active." << std::endl;
+            thread_states.resize(max_threads);
+        }
+    }
+
+    // Register the tracer in each thread.
+    for (size_t i = 0; i < thread_states.size(); i++) {
+        PyThreadState* thread_state = thread_states[i];
+        PyThreadState_Swap(thread_state);
+
+        auto ctx = (TraceContext*) TraceContextType.tp_alloc(&TraceContextType, 0);
+        ctx->thread_id_ = (uint8_t)i;
+        ctx->thread_state_ = thread_state;
+        ctx->initial_us_ = t0;
+        trace_contexts_.push_back(ctx);
+
+        // When we begin profiling there are already frames on the Python
+        // interpreter stack. To ensure a complete trace, we must push calls
+        // to all the prior frames onto our event stack. (We stop at depth=128)
+        std::vector<PyFrameObject*> current_stack;
+        auto frame = PyEval_GetFrame();
+        size_t depth = 0;
+        while (frame != nullptr && depth <= 128) {
+            current_stack.push_back(frame);
+            frame = frame->f_back;
+            depth++;
+        }
+        for (auto it = current_stack.rbegin(); it != current_stack.rend(); it++) {
+            record_py_call(ctx, *it);
+        }
+
+        // Note:
+        //   This profile will not compose with other profilers, and cannot be
+        //   round tripped via `sys.settrace(sys.gettrace())`
```

Comment:
What does this comment refer to?

Reply:
Other CPython profilers (e.g. `coverage`) that also use `PyEval_SetProfile`. I changed the language to `This profile will not compose with other CPython profilers`